### PR TITLE
[Tokenomics] Preparation for proportional relay mining rewards based on difficulty

### PR DESCRIPTION
**Pull request by Olshansk**
## Summary

No real business logic changes, but preparation + tests + helpesr + TODOs for #781.

## Issue

- Preparation for #781
- Pre-requisite for #818

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [x] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Documentation**: `make docusaurus_start`; only needed if you make doc changes
- [ ] **Unit Tests**: `make go_develop_and_test`
- [ ] **LocalNet E2E Tests**: `make test_e2e`
- [ ] **DevNet E2E Tests**: Add the `devnet-test-e2e` label to the PR.

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have commented my code
- [ ] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

### CODE REVIEW

File: pkg/crypto/protocol/relay_difficulty_test.go [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+		numActualLowVolumeSvcRelays++
+		if numApplicableLowVolumeSvcRelays >= int(targetNumRelays) {
+			break
+		}
+	}
+	numEstimatedLowVolumeSvcRelays := float64(numApplicableLowVolumeSvcRelays) * lowVolumeRelayMultiplier
+	fractionLowVolumeSvcRelays := float64(numApplicableLowVolumeSvcRelays) / float64(numActualLowVolumeSvcRelays)
+
+	// Ensure probabilities of a relay being applicable is within the allowable delta
+	require.InDelta(t, highVolumeRelayProbability, fractionHighVolumeSvcRelays, allowableDelta*highVolumeRelayProbability)
```
Comment (red-0ne) [inline pkg/crypto/protocol/relay_difficulty_test.go:359]: Magic! :mage_man:

**Review by red-0ne — COMMENTED** [review]
Made a second pass to better understand and effectively review this PR (esp. `relay_difficulty_test.go` which explains all that is happening).

This is great! this is the foundation of probabilistic relay estimation. Hats off :tophat:

File: x/tokenomics/keeper/update_relay_mining_difficulty_test.go [hunk 2]
```diff
@@ -2,20 +2,93 @@ package keeper_test
...
+		svcRelayMiningDifficulty, found := keeper.GetRelayMiningDifficulty(ctx, svcId)
+		require.True(t, found)
+
+		// Since the num relays is monotonically increasing, the EMA should also
+		// be increasing but always less than the num relays.
+		require.Greater(t, numRelays, prevEMA)
+		require.Greater(t, svcRelayMiningDifficulty.NumRelaysEma, prevEMA)
+		prevEMA = svcRelayMiningDifficulty.NumRelaysEma
+
+		// Only enforce that the target hash is monotonically decreasing if it is not the default
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply x/tokenomics/keeper/update_relay_mining_difficulty_test.go:42]: That's the goal: expontential moving average.

File: x/tokenomics/keeper/update_relay_mining_difficulty.go [hunk 3]
```diff
@@ -192,16 +134,22 @@ func newDefaultRelayMiningDifficulty(
 	serviceId string,
 	numRelays uint64,
 ) tokenomicstypes.RelayMiningDifficulty {
-	logger = logger.With("helper", "newDefaultRelayMiningDifficulty")
+	logger = logger.With("helper", "NewDefaultRelayMiningDifficulty")
+
+	// Compute the target hash based on the number of relays seen for the first time.
+	newDifficultyHash := protocol.ComputeNewDifficultyTargetHash(prooftypes.DefaultRelayDifficultyTargetHash, TargetNumRelays, numRelays)
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply x/tokenomics/keeper/update_relay_mining_difficulty.go:140]: Use `protocol.BaseRelayDifficultyHashBz` instead. I've updated the comment as well.

File: x/proof/keeper/proof_validation.go [hunk 4]
```diff
@@ -41,8 +41,6 @@ import (
 )
 
 // EnsureValidProof validates the proof submitted by the supplier is correct with
-// respect to an on-chain claim.
-//
```
Comment (red-0ne) [inline x/proof/keeper/proof_validation.go:45]: This was probably truncated by accident
```
// EnsureValidProof validates the proof submitted by the supplier is correct with
// respect to an on-chain claim.
//
```

**Review by red-0ne — APPROVED** [review]
Love how this is taking form!

Left a small NIT but preemptively approving.

**Comment by red-0ne** [comment]
Regarding this test error
image

Adding a `time.Sleep` before event filtering should work around the issue

**Review by red-0ne — APPROVED** [review]

**Comment by Olshansk** [comment]
Commenting out a single test that seems to be related to an architecture floating point issue that @red-0ne will addres..

Screenshot 2024-09-20 at 4 55 09 PM

Screenshot 2024-09-20 at 5 12 13 PM

**Pull request merged**